noisy-simulator/src/state_vector_simulator.rs

```diff
@@ -0,0 +1,359 @@
+            let p = norm_squared / renormalization_factor;
+            summed_probability += p;
+            if p >= TOLERANCE {
+                last_non_zero_probability = p;
+                last_non_zero_probability_index = i;
+                if summed_probability > random_sample {
+                    self.data = state_copy;
+                    self.renormalize_with_norm_squared(norm_squared)?;
+                    return Ok(());
+                }
+            }
+        }
+        if summed_probability + TOLERANCE > random_sample && last_non_zero_probability >= TOLERANCE
+        {
+            return Err(Error::FailedToSampleKrausOperators);
+        }
+        apply_kernel(
+            &mut self.data,
+            &kraus_operators[last_non_zero_probability_index],
+            qubits,
+        )?;
+        self.renormalize()
+    }
+}
+
+/// A quantum circuit simulator using a state vector.
+pub struct StateVectorSimulator {
+    /// A `StateVector` representing the current state of the quantum system.
+    state: Result<StateVector, Error>,
+    /// Dimension of the density matrix. We need this field to verify the size of the
+    /// quantum system in the `set_state` method in the case that `self.state == Err(...)`.
+    dimension: usize,
+    /// Random number generator used for probabilistic operations.
+    rng: StdRng,
+}
+
+impl StateVectorSimulator {
+    fn check_out_of_bounds_qubits(&self, qubits: &[usize]) -> Result<(), Error> {
+        let number_of_qubits = self.state.as_ref()?.number_of_qubits;
+        if let Some(id) = qubits.iter().find(|id| **id >= number_of_qubits) {
+            Err(Error::QubitIdOutOfBounds(*id))
+        } else {
+            Ok(())
+        }
+    }
+}
+
+impl NoisySimulator for StateVectorSimulator {
+    type State = StateVector;
+
+    /// Creates a new `TrajectorySimulator`.
+    fn new(number_of_qubits: usize) -> Self {
+        let state_vector = StateVector::new(number_of_qubits);
+        let dimension = state_vector.dimension();
+        Self {
+            state: Ok(state_vector),
+            dimension,
+            rng: StdRng::from_entropy(),
+        }
+    }
+
+    fn set_rng_seed(&mut self, seed: u64) {
+        self.rng = StdRng::seed_from_u64(seed);
+    }
+
+    /// Apply an operation to given qubit ids.
+    fn apply_operation(&mut self, operation: &Operation, qubits: &[usize]) -> Result<(), Error> {
+        self.check_out_of_bounds_qubits(qubits)?;
+
+        let renormalization_factor = self
+            .state
+            .as_mut()?
+            .effect_probability(operation.effect_matrix(), qubits)?;
+        self.state.as_mut()?.trace_change *= renormalization_factor;
+
+        if let Err(err) = self.state.as_mut()?.sample_kraus_operators(
+            operation.kraus_operators(),
+            qubits,
+            renormalization_factor,
+            self.rng.gen(),
+        ) {
+            handle_error!(self, err);
+        };
+
+        Ok(())
```

Comment:
Mentioned in another comment, but I would have expected map_err to handle this.

Reply:
Explained in the other comment.